cpp/src/arrow/compute/exec/spilling_join.cc

```diff
@@ -0,0 +1,351 @@
+            // the final Bloom filter
+            partition_idx_ = (bloom_filter_ == nullptr)
+                ? partition_idx_ = SpillingAccumulationQueue::kNumPartitions
+                : build_accumulator_.hash_cursor();
+            return BuildNextBloomFilter(thread_index);
+        }
+
+        Status SpillingHashJoin::PushBloomFilterBatch(size_t thread_index, int64_t batch_id)
+        {
+            const uint64_t *hashes = build_accumulator_.GetHashes(
+                partition_idx_,
+                static_cast<size_t>(batch_id));
+            size_t num_rows = build_accumulator_.row_count(
+                partition_idx_,
+                static_cast<size_t>(batch_id));
+            return builder_->PushNextBatch(
+                thread_index,
+                static_cast<int64_t>(num_rows),
+                hashes);
+        }
+
+        Status SpillingHashJoin::BuildNextBloomFilter(size_t thread_index)
+        {
+            if(partition_idx_ >= SpillingAccumulationQueue::kNumPartitions)
+                return OnBloomFilterFinished(thread_index);
+
+            size_t num_rows = build_accumulator_.row_count(partition_idx_);
+            size_t num_batches = build_accumulator_.batch_count(partition_idx_);
+
+            // partition_idx_ is incremented in the callback for the taskgroup
+            bloom_filter_->partitions[partition_idx_] =
+                arrow::internal::make_unique<BlockedBloomFilter>();
+
+            RETURN_NOT_OK(builder_->Begin(
+                              num_threads_,
+                              ctx_->cpu_info()->hardware_flags(),
+                              ctx_->memory_pool(),
+                              num_rows,
+                              num_batches,
+                              bloom_filter_->partitions[partition_idx_].get()));
+                
+            return callbacks_.start_task_group(
+                task_group_bloom_[partition_idx_],
+                build_accumulator_.batch_count(partition_idx_));
+        }
+        
+        Status SpillingHashJoin::OnBloomFilterFinished(size_t thread_index)
+        {
+            RETURN_NOT_OK(callbacks_.bloom_filter_finished(thread_index));
+            bool expected = false;
+            if(!bloom_or_probe_finished_.compare_exchange_strong(expected, true))
+                return StartCollocatedJoins(thread_index);
+            return Status::OK();
+        }
+
+        Status SpillingHashJoin::OnBloomFiltersReceived(size_t thread_index)
+        {
+            bloom_ready_.store(true, std::memory_order_release);
+            return Status::OK();
+        }
+
+        Status SpillingHashJoin::OnProbeSideBatch(size_t thread_index, ExecBatch batch)
+        {
+            if(bloom_ready_.load())
+            {
+                RETURN_NOT_OK(callbacks_.apply_bloom_filter(
+                    thread_index,
+                    &batch));
+            }
+            return probe_accumulator_.InsertBatch(
```

Comment:
What if we inserted a batch into the queue before Bloom filter was ready? Will we apply the Bloom filter later on when we read a batch from the queue?

Reply:
Made it so that we apply Bloom filter when we read it back from disk.